src/lammpsio/snapshot.py

```diff
@@ -170,37 +258,84 @@ def to_hoomd_gsd(self, type_map=None):
             frame.particles.image = self.image.copy()
         if self.has_typeid():
             frame.particles.typeid = numpy.zeros(self.N, dtype=int)
-            if type_map is None:
-                sorted_typeids = numpy.sort(numpy.unique(self.typeid))
-                frame.particles.types = [str(typeid) for typeid in sorted_typeids]
-                for typeidx, typeid in enumerate(sorted_typeids):
-                    frame.particles.typeid[self.typeid == typeid] = typeidx
+            if type_map is not None:
+                warnings.warn(
+                    "type_map is deprecated, use label instead.",
+                    DeprecationWarning,
+                    stacklevel=2,
+                )
+                type_label_map = LabelMap(map=type_map)
             else:
-                frame.particles.types = list(type_map.values())
-                reverse_type_map = {
-                    typeid: typeidx for typeidx, typeid in enumerate(type_map.keys())
-                }
-                for i, typeid in enumerate(self.typeid):
-                    frame.particles.typeid[i] = reverse_type_map[typeid]
+                type_label_map = self.type_label
+            if self.has_typeid():
+                frame.particles.typeid = numpy.zeros(self.N, dtype=int)
+                type_label_map = _set_type_id(
+                    self.typeid, frame.particles.typeid, type_label_map
+                )
+            frame.particles.types = type_label_map.types
         if self.has_charge():
             frame.particles.charge = self.charge.copy()
         if self.has_mass():
             frame.particles.mass = self.mass.copy()
         if self.has_molecule():
             frame.particles.body = self.molecule - 1
+        if self.has_bonds():
```

Comment:
I forget the details of this method, but iirc, I think this should be:
```suggestion
        if self.bonds is not None:
```
because `has_bonds()` also checks for N > 0, but it is allowed to be 0 now.

Reply:
I think gsd.hoomd's default behavior is for `frame.bonds.N` to be zero if undefined for all the topology. 

```python
# make a GSD frame
frame = gsd.hoomd.Frame()

frame.configuration.step = 3
frame.configuration.box = [4, 5, 6, 0.1, 0.2, 0.3]
frame.particles.N = 2
frame.particles.position = [[0.1, 0.2, 0.3], [-0.1, -0.2, -0.3]]
frame.particles.image = [[1, -1, 0], [0, 2, -2]]
frame.particles.velocity = [[1, 2, 3], [-4, -5, -6]]
frame.particles.body = [0, -1]
frame.particles.types = ["A", "B"]
frame.particles.typeid = [1, 0]
frame.particles.mass = [3, 2]
frame.particles.charge = [-1, 1]

print(frame.bonds.N)
>>> 0 
```

I think this means we should change the `from_hoomd_gsd` to be 
```python
if (
            frame.bonds.N != 0
            or frame.bonds.group is not None
            or frame.bonds.typeid is not None
            or frame.bonds.types is not None
        ):
````
and then using `self.bonds is not None` would work here

Reply:
Yeah, that seems reasonable if the default value for that is 0 and not None! If it defaults to 0, you can also remove the bit about `is not None` when creating the topology objects.

Reply:
> you can also remove the bit about is not None when creating the topology objects.

Reply:
This part will always evaluate true now:
```
N=frame.bonds.N if frame.bonds.N is not None else 0
```
so it can be
```
N=frame.bonds.N
```